Generate DBLayer tests with quickcheck-state-machine

rvl opened the pull request:
Relates to #154.

# Overview

We can generate our DBLayer tests with quickcheck.

I used [Edsko's blog post](https://iohk.io/blog/an-in-depth-look-at-quickcheck-state-machine/) as a guide.

# Comments

Todo list:

- [x] DBLayer methods `put/readPrivateKey`
- [x] shrinking of TxHistory doesn't work
- [x] Some more interesting tests can be tagged
- [x] Testing of the SeqState wallet
- [x] Tweak weightings in the command generator.
- [x] ~Use QSM parallel testing~
- [x] Run QSM tests on Sqlite ~and MVar~ DBLayers.

KtorZ commented:
@uroboros We likely won't. This is really just to facilitate migration of wallets from previous versions. The problem doesn't really occur (or is actually completely different) for sequential derivation because addresses are actually derived .. in sequence. So, as a user, one'll bump into the same addresses on a fresh new wallet. 

The big issue with the sequential derivation for exchanges at the moment is the number of unused addresses they need to have in supply. Having a `gap` of 100 may not be enough for big exchanges. But having a gap of 1000+ may be extremely inefficient to manipulate in the wallet backend. We are investigating at the moment how they deal with wallets in other blockchain that also follow a sequential derivation scheme.

deepfire commented:
@KtorZ, it would be really nice to get some test coverage for this.

Thinking about use cases an exchange will have (primarily collisions with existing ones?).

deepfire commented:
@KtorZ, one thing is that we'll need this in 1.4 as well.

deepfire commented:
@KtorZ, no, 1.4, because customers..

parsonsmatt commented on src/Cardano/Wallet/API/V1/Addresses.hs:
```diff
@@ -19,4 +19,8 @@ type API = Tags '["Addresses"] :>
                        :> Get '[ValidJSON] (APIResponse WalletAddress)
+      :<|> "wallets" :> CaptureWalletId :> "accounts" :> CaptureAccountId :> "addresses"
+        :> Summary "Batch import existing addresses"
+        :> ReqBody '[ValidJSON] [WalAddress]
```
So this *is* going to perform poorly on large accounts. The "right" way to do this is with a file upload (ie [Multipart](https://haskell-servant.readthedocs.io/en/stable/cookbook/file-upload/FileUpload.html)) to a `tmp` directory we can stream from. Then we can process this in batches/chunks in constant memory.

How many addresses are exchanges wanting to backup and restore? We should find this out and determine if it's a realistic bottleneck -- 100k addresses is 160MB of memory just in list overhead, but 1k is only 1.6MB list overhead.

KtorZ commented:
@matt.parsons agreed, hence the warning in the documentation and the comments in the PR / corresponding ticket.
I went for the naive and fastest approach here because:

- We basically need this for Friday
- That's most likely something exchanges will use one time only when they move from 1.3 to 1.4
- This came up as an unexpected and unplanned requirement, so we deliver whatever we can in the allowed time frame.

Performances were the least concern here and if we intend to keep on supporting this feature, I 100% agree with what you just wrote.

rvl committed:
Generate DBLayer tests with quickcheck-state-machine

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DBSpec2.hs:
```diff
@@ -0,0 +1,595 @@
+mReadPrivateKey :: MWid -> MockOp (Maybe MPrivKey)
+mReadPrivateKey wid m@(M cp _ _ pk _)
+    | wid `Map.member` cp = (Right (Map.lookup wid pk), m)
+    | otherwise           = (Left (NoSuchWallet wid), m)
```
Wouldn't it be possible to leverage our already existing `MVar` implementation as the Mock, instead of re-encoding everything here :thinking: ? We've already some good unit tests which make sure that the MVar implementation works rather okay.

KtorZ reviewed (commented):
Interesting :thinking: .. I got to re-read Edsko's blog post on the matter. One thing that bother me is that I couldn't find where pre-conditions for commands are located in this implementation. I wouldn't expect a "removeWallet" to be a valid transition if there hasn't been any `createWallet` before...
so, I might be missing something from _what_ is being tested here maybe?

paweljakubas commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -104,6 +104,7 @@ test-suite unit
     , aeson
     , aeson-qq
     , async
+    , bifunctors
```
use `base` bifunctors

rvl replied:
I tried to remove it, but this package contains `Data.Bifunctor.TH.deriveBifunctor`, which I don't think `base` provides.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/TestWallet.hs:
```diff
@@ -0,0 +1,68 @@
+import Test.QuickCheck
+    ( Arbitrary (..) )
+
+import qualified Data.ByteString.Char8 as B8
```
import should comply with our standards

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/TestWallet.hs:
```diff
@@ -0,0 +1,68 @@
+instance TxId DummyTarget where
+    txId = Hash . B8.pack . show
+
+newtype DummyState = DummyState Int
```
`DummyState` probably should be like : 
```
newtype DummyStateSqlite = DummyStateSqlite (UTxO, PendingTx)
    deriving (Show, Eq, NFData)
```
Or even better to prepare test machine to take any state, ie. use state type parameter and then check this both in `MVarSpec` and `SqliteSpec`

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DBSpec2.hs:
```diff
@@ -0,0 +1,655 @@
+import Prelude hiding
+    ( elem )
+
+import Control.Monad.Trans.Except
```
imports refactoring is needed

akegalj commented on lib/core/test/unit/Cardano/Wallet/DBSpec2.hs:
```diff
@@ -0,0 +1,655 @@
+  Mock implementation
+-------------------------------------------------------------------------------}
+
+-- fixme: remove MWid silliness
```
what for is `MWid` ?

akegalj replied:
aha - I guess its a Mock wrapper

rvl replied:
MWid allows using mock wallet IDs like a,b,c instead of the real types which have to be a blake2b digests.

akegalj commented on lib/core/test/unit/Cardano/Wallet/DBSpec2.hs:
```diff
@@ -0,0 +1,655 @@
+    deriving (Show, Generic)
+
+emptyMock :: Mock
+emptyMock = M Map.empty Map.empty Map.empty Map.empty (MConn 0)
```
I guess `mempty` could work as well instead of `Map.empty`

akegalj reviewed (commented):
I have to admit that I didn't understand around 50% of the mechanics here on a first skim although I was using this lib a year ago. I guess edsko did use it in a more advanced way than I did.

Would have to skim it once again after I read edsko's blog post

rvl committed:
Sqlite: Make listWallets ordering explicit

QSM test was failing with this:

Open
CreateWallet b
CreateWallet a
ListWallets

PostconditionFailed "PredicateC (Resp (Right (WalletIds [MWid \"b\",MWid \"a\"])) :/= Resp (Right (WalletIds [MWid \"a\",MWid \"b\"])))" /= Ok

rvl committed:
Sqlite: Transaction inputs need to be ordered

rvl committed:
Move cleanDB to DBLayer

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DBSpec2.hs:
```diff
@@ -797,17 +701,18 @@ instance ToExpr MWid where
 spec = describe "DBLayer state machine tests" $ do
     it "Sequential tests" $
-        quickCheck (prop_sequential Nothing)
+        quickCheck . prop_sequential
```
By the by @rvl, be careful with the use of `quickCheck`. quickCheck doesn't report errors with an error result or failure exit code. So, if the property fails, it just prints a counter example, but `hspec` doesn't report the test as fail. 
Instead, use `property` from QuickCheck and let hspec instrument the run of the property such that it can collects the result and properly report a failure if any!

rvl replied:
Fixed, thanks.

KtorZ committed:
run sqlite migration silently

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -0,0 +1,944 @@
+mPutTxHistory wid txs' m@(M cp metas txs pk)
+    | wid `Map.member` cp = (Right (), M cp metas (Map.alter appendTxs wid txs) pk)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+    where appendTxs = Just . (<> txs') . fromMaybe mempty
```
Map union is _left-biased_, which means that in case of conflicting keys, the entries from the left map will prevail. However, we've given another semantic to our database-layer: new transaction should overwrite existing ones (debatable, see below). I wondered however why the tests failed to catch that as I remember making sure to generate transaction ids with a rather small entropy. 

Yet, the `Arbitrary` instance for `Hash "Tx"` has been changed to generate 32-byte hashes across the whole range, which makes 16^32 (2^128) different hashes, and therefore, we probably never generate twice the same tx id.

Changing this to something that generates more conflict indeed yield the following error:

```hs
       uncaught exception: SqliteException
       SQLite3 returned ErrorConstraint while attempting to perform step: UNIQUE constraint failed: tx_out.tx_id, tx_out.index
       (after 89 tests and 189 shrinks)
```

Now the question about whether the semantic we initially gave is correct is a valid question. The rationale was that, during rollbacks, we will have to possibly replace metadata of some existing transactions. One approach could be to overwrite transactions with some others in order to avoid having to perform a delete and then an insert. However, there's some values in making sure we correctly clean-up the database, throwing an exception if we try to insert twice a transaction -- this would clearly alert us that something's broken in our cleanup process.

What do you think @rvl ? Also, if we keep the generator as it is, it may be worth adding a precondition to verify that we don't already know of a transaction with a newly generated id. Conflicts are very unlikely over 2^128, but not impossible provided we generate quite many transactions on quite many runs...

KtorZ reviewed (commented):
Getting some failures while running the parallel tests. 

<details>
    <summary>SQLite Parallel failure</summary>

```
       uncaught exception: SqliteException
       SQLite3 returned ErrorError while attempting to perform step: cannot start a transaction within a transaction
       (after 5 tests and 2 shrinks)
         ParallelCommands
           { prefix = Commands { unCommands = [] }
           , suffixes =
               [ Pair
                   { proj1 =
                       Commands
                         { unCommands =
                             [ Command
                                 (At (ReadPrivateKey (Val (MWid "c"))))
                                 (At (Resp (Right (PrivateKey Nothing))))
                                 []
                             ]
                         }
                   , proj2 =
                       Commands
                         { unCommands =
                             [ Command
                                 (At
                                    (PutWalletMeta
                                       (Val (MWid "c"))
                                       WalletMetadata
                                         { name = WalletName { getWalletName = "bulbazaur" }
                                         , passphraseInfo =
                                             Just
                                               WalletPassphraseInfo
                                                 { lastUpdatedAt =
                                                     (1864 - 05 - 11) (22 : 24 : 03.767284750574) UTC
                                                 }
                                         , status = Ready
                                         , delegation = NotDelegating
                                         }))
                                 (At (Resp (Left (NoSuchWallet (Reference (Symbolic (Var 0)))))))
                                 [ Var 0 ]
                             ]
                         }
                   }
               ]
           }
```
</details>

<details>
    <summary>SQLite parallel failure (2)</summary>

```
uncaught exception: SqliteException
 SQLite3 returned ErrorError while attempting to perform step: not an error
 (after 4 tests)
   ParallelCommands
     { prefix = Commands { unCommands = [] }
     , suffixes =
         [ Pair
             { proj1 =
                 Commands
                   { unCommands =
                       [ Command
                           (At
                              (CreateWallet
                                 (MWid "c")
                                 (Wallet
                                    UTxO { getUTxO = fromList [] }
                                    (fromList [])
                                    SlotId { epochNumber = 0 , slotNumber = 0 }
                                    SeqState
                                      { internalPool =
                                          AddressPool
                                            { accountPubKey =
                                                Key
                                                  { getKey =
                                                      XPub
                                                        { xpubPublicKey =
                                                            "Q\170\GS\202\198\&2KA\203\CANN'X\154 \139\DEL\FS\148\FSb\SO\RS\r\DLEAL\151\153\137\167\194"
                                                        , xpubChaincode =
                                                            ChainCode
                                                              "\204\196\"I\225y\132\196L\243\128\180\137\246,W\248@\137\225P$[\244\156Cm\v\151\t\197\143"
                                                        }
                                                  }
                                            , gap = AddressPoolGap { getAddressPoolGap = 15 }
                                            , indexedAddresses =
                                                fromList [ ... ]
                                            }
                                      , externalPool =
                                          AddressPool
                                            { accountPubKey =
                                                Key
                                                  { getKey =
                                                      XPub
                                                        { xpubPublicKey =
                                                            "Q\170\GS\202\198\&2KA\203\CANN'X\154 \139\DEL\FS\148\FSb\SO\RS\r\DLEAL\151\153\137\167\194"
                                                        , xpubChaincode =
                                                            ChainCode
                                                              "\204\196\"I\225y\132\196L\243\128\180\137\246,W\248@\137\225P$[\244\156Cm\v\151\t\197\143"
                                                        }
                                                  }
                                            , gap = AddressPoolGap { getAddressPoolGap = 10 }
                                            , indexedAddresses =
                                                fromList [ ... ]
                                            }
                                      , pendingChangeIxs = PendingIxs { pendingIxsToList = [] }
                                      })
                                 WalletMetadata
                                   { name = WalletName { getWalletName = "bulbazaur" }
                                   , passphraseInfo =
                                       Just
                                         WalletPassphraseInfo
                                           { lastUpdatedAt =
                                               (1864 - 05 - 06) (09 : 58 : 17.690387360868) UTC
                                           }
                                   , status = Restoring (Quantity Percentage { getPercentage = 66 })
                                   , delegation = NotDelegating
                                   }))
                           (At (Resp (Right (NewWallet (Reference (Symbolic (Var 0)))))))
                           [ Var 0 ]
                       ]
                   }
             , proj2 =
                 Commands
                   { unCommands =
                       [ Command
                           (At (ReadWalletMeta (Val (MWid "b"))))
                           (At (Resp (Right (Metadata Nothing))))
                           []
                       ]
                   }
             }
         ]
     }
```
</details>

The parallel MVar tests are also failing sometimes:

<details>
   <summary>MVar parallel failure</summary>

```
  1) Cardano.Wallet.DB.MVar, MVar state machine tests, Parallel
       *** Failed! (after 95 tests and 203 shrinks):
       Exception:
         insertConcrets: impossible.
         CallStack (from HasCallStack):
           error, called at src/Test/StateMachine/Types/Environment.hs:73:3 in quickcheck-state-machine-0.6.0-w8NmrpYN7DCoqRW8X5Sia:Test.StateMachine.Types.Environment
```
</details>

I am not sure whether this is due to our implementation, SQLite itself, or, due to the QSM library... I remember Edsko saying quite a few times that the parallel testing wasn't really ready for serious uses... 

Beside, on parallel tests, some pre-conditions are rather hard to satisfy / meet. I believe this is / ought to be handled properly by the state-machine library. If I recall correctly, the theory behind doing parallel QSM testing is like running every possible interleaves of the commands sequentially. But for instance, what if you create a wallet in one worker, and simply query it with the other worker. If the first worker runs first, the second end up with a wallet, if the second runs first, it got nothing. 

To be honest, I'd rather have us focusing on the sequential one in this PR to get moving already, instead of a never-ending stream of work which adds always more and more complexity. Let's try to keep the scope of PR well-defined and open new PRs for additions that are out-of-scope.

rvl committed:
Disable QSM parallel tests until we're sure they work

rvl committed:
QSM: Adjust the putTxHistory model to handle updating transactions

And put back the Arbitrary (Hash "Tx") instance.

rvl committed:
Sqlite: Allow adding the same transaction with putTxHistory

rvl committed:
SqliteSpec: Skip TxHistory put+read tests for Sqlite

The model of dbPropertyTests doesn't match how the relational database
works.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -213,7 +215,8 @@ newDBLayer fp = do
         , listWallets = runQuery' $
-              map (PrimaryKey . unWalletKey) <$> selectKeysList [] []
+              map (PrimaryKey . unWalletKey) <$>
+              selectKeysList [] [Asc WalTableId]
```
Just wondering: is there any particular reason for this ordering?

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -0,0 +1,962 @@
+-------------------------------------------------------------------------------}
+
+newtype At f r
+    = At (f (Reference WalletId r))
```
Would it be possible to add some kind of comment to describe what this type is for? For example, what do **`f`** and **`r`** refer to?

rvl committed:
DBLayer QSM: Remove WalletId expressions

There is no complex dependency between DBLayer commands, so the Expr
type is not needed.

rvl committed:
DBSpec: Fix comment on Arbitrary TxIn

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -692,11 +699,11 @@ selectSeqStatePendingIxs ssid =
+    insertState (wid, sl) _ = insert_ (SeqState wid sl)
+    selectState (wid, sl) = fmap (const DummyState) <$>
+        selectFirst [SeqStateTableWalletId ==. wid, SeqStateTableCheckpointSlot ==. sl] []
+    deleteState wid = deleteWhere [SeqStateTableWalletId ==. wid]
```
This shouldn't be in the source code IMO :/

KtorZ commented on lib/core/src/Cardano/Wallet/DB.hs:
```diff
@@ -146,3 +147,7 @@ newtype ErrWalletAlreadyExists
+
+-- | Clean a database by removing all wallets.
+cleanDB :: Monad m => DBLayer m s t -> m ()
+cleanDB db = listWallets db >>= mapM_ (runExceptT . removeWallet db)
```
I am not sure we do want to keep this in the source code. This is purely for testing.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DBSpec.hs:
```diff
@@ -639,10 +652,16 @@ prop_parallelPut putOp readOp resolve dbLayer (KeyValPairs pairs) =
+data SkipTests
+    = RunAllTests
+    | SkipTxHistoryReplaceTest
+    deriving (Show, Eq)
```
This looks a bit dodgy :no_mouth: . What exactly is the issue with this test? If there's an issue, let's disable it temporarily and open a bug ticket about fixing it instead?

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -0,0 +1,907 @@
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.DB.StateMachine
```
I just realized Edsko's blog post isn't mentioned anywhere in this file. It'll be nice to have a reference to it as a comment :+1:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -0,0 +1,907 @@
+unMockWid (MWid wid) = WalletId . hash . B8.pack $ wid
+
+-- | Represent (XPrv, Hash) as a string.
+type MPrivKey = String
```
I wonder why we do have this mock `XPrv` and `Hash` in a first place? The generator only takes keys among a fixed list of 3 elements anyway, so we could as well just go for the right type from the start :thinking:  ?

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -0,0 +1,907 @@
+
+-- | Mock wallet ID -- simple and easy to read
+newtype MWid = MWid String
+    deriving (Show, Eq, Ord, Generic)
```
Same remark here, I don't think the extra indirection is helpful here. It adds another layer of indirection that has  little value? Arbitrary wallet id are easy to generate (especially if just taken from a fixed list of elements) and, this does increase the cognitive load necessary to process this module. Don't you think?

KtorZ committed:
re-enable prop_sequential_put for tx history

Problem here was that our generator would generate inputs, outputs and
transaction ids independently. While this didn't matter much with the
MVar implementation, it does create weird inconsistency with the
relational model because indeed, we expect inputs associated with a
particular tx id to not differ when associated with the same tx id.

Since a transaction id is actually a hash over the content of the
transaction, changing the transaction should yield a different id.
So now, we generate "fake" txid by computing them from the actual
transaction content (as done in production, we simply don't hash
it in the test as this has little value here but cost quite some
computing power).

KtorZ committed:
Call 'cleanDB' before each property run.

We use to create a fresh new DB layer before the run of each property
because this was very cheap. Now that we do create a db layer only once
before a given property, we need to make sure that the state of the DB
is at least cleaned up between each property.

KtorZ committed:
remove redundant constraints

KtorZ reviewed (approved):
Merging as-is. We can iterate on a few remaining parts in future PRs.

Nice job on pulling that out @rvl 
@jonathanknowles Feedback is still welcome afterwards.

I've also ran the sequential tests ~10000 times trying to catch something. Seems pretty stable now.
